### ISSUE TITLE
Failed connection attempts are not released from connection pool

### ISSUE THREAD

**Issue by yamnigh** [gajus/slonik#631]
Discovered while db was down for upgrade.  Even though db came back online, app was unable to create new connections.

## Expected Behavior
When a connection attempt fails, it should be removed from the connection pool, so new connections can be attempted later

## Current Behavior
When a connection attempt fails, it is not removed from the connection pool; pool eventually fills up with empty connections, causing infinite queues for new connections

## Possible Solution
1. Need to be sure failed connection attempts don't hold onto pool connections
2. Could close failed connections if connectionTimeout is reached

## Steps to Reproduce
See demo repo
https://github.com/yamnigh/demo-slonik-issue-2024-connection-failure

Tested past versions, found issue began in `40.1.2`, previous version`40.1.1`;  can see diff [here](https://github.com/gajus/slonik/compare/v40.1.1...v40.1.2)

tldr
1. Create pool to DB
2. Turn off DB
3. Create new connections until pool is full
4. See new connections are queued indefinitely

**Comment by gajus** [comment]
ah, interesting. I am fairly sure I've encountered this in real-world. Thanks for the patch!
